torch/lib/c10d/ProcessGroup.cpp

```diff
@@ -56,6 +56,15 @@ void ProcessGroup::Work::finish(std::exception_ptr exception) {
   cv_.notify_all();
 }
 
+void ProcessGroup::Work::finishAndThrow(std::exception_ptr exception) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  completed_ = true;
+  exception_ = exception;
```

Comment:
Could nullptr `exception `overwrite` exception_`, hence we miss the previous exception?

Reply:
Yes, this could happen, but this is possible in the gloo case - if no exception is caught by `waitSend/waitRecv`, then `exception_` is set to `nullptr` and nothing is thrown, even if an exception was previously set.

We can remove the `nullptr` from being the default value for the `exception_ptr` argument, so the user has to explicitly pass in a `nullptr` if they want to overwrite the last exception.